src/docfx/lib/schema/JsonSchemaValidationContext.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal class JsonSchemaValidationContext
+    {
+        private readonly JsonSchema _root;
+        private readonly Dictionary<string, JsonSchema> _definitions;
+
+        public JsonSchemaValidationContext(JsonSchema root)
+        {
+            Debug.Assert(root != null);
+
+            _root = root;
+            _definitions = _root.Definitions.ToDictionary(k => $"#/definitions/{k.Key}", v => v.Value);
+            _definitions.Add("#", root);
+        }
+
+        public JsonSchema GetDefinition(JsonSchema jsonSchema, HashSet<string> recursions = null, string key = "#")
```

Comment:
consider simplify public interface make recursion parameters private. It might be easier to pass `ref` to convey that we are getting a schema from a $ref

```suggestion
        public JsonSchema GetDefinition(string ref)
```

Reply:
We need the whole schema